## Will Paz

In [1]:
import numpy as np
import math as mth

### Part 1: 

Write out by hand  the 5x5 stochastic matrix describing the random walk. Using this matrix as a guide make a function rw_matrix(n,p) that outputs the nxn stochastic matrix describing the random walk.

In [2]:
def rw_matrix(n,p):
    """
    Generates a the nxn stochastic matrix describing the random walk
    Inputs: n, the size of the square matrix, and p, the desired probability
    Output: nxn Stochastic matrix
    """
    S = np.zeros((n,n))
    
    for i in range(n):
        S[i,i] = p  
        
    for i in range(n-1):
        S[i+1,i] = (1-p)/2
        S[i,i+1] = (1-p)/2
        
    S[0,n-1] = (1-p)/2
    S[n-1,0] = (1-p)/2
    
    lambda2 = p+(1-p)*mth.cos(2*mth.pi/n)
            
    return S, lambda2

In [3]:
A,L = rw_matrix(5,(1/3))

In [4]:
E,V = np.linalg.eigh(A)

In [5]:
E[3]

0.539344662916632

In [6]:
L

0.5393446629166316

### Checking the 2nd Eigenvalues

With n = 5 and p = 1/3, and using rw_matrix(5,1/3) we can get the results below.

Using np.linalg.eigh(A), the second eigenvalue = 0.539344662916632

Using a discrete Fourier transform, the value = 0.5393446629166316

So, yes, they are equal (enough).

### Random Walks with n = 11 and p = 0.15 (walker starts at node 1/index 0)

In [7]:
B,L2 = rw_matrix(11,.15)

#### 1. What is the probability that the walker is at node 6 after 3 steps?

In [8]:
cubed = np.linalg.matrix_power(B, 3)

In [9]:
cubed[0,5]

0.0

Answer: 0.0

#### 2. What is the probability that the walker is at node 2 after 3 steps?

In [10]:
cubed[0,1]

0.258984375

Answer: 0.258984375

#### 3.  What is the probability that the walker is at node 6 after 10 steps?

In [11]:
tened = np.linalg.matrix_power(B, 10)

In [12]:
tened[0,5]

0.04969394382680024

Answer: 0.04969394382680024

#### 4.  What is the probability that the walker is at node 2 after 10 steps?

In [13]:
tened[0,1]

0.12389366636436545

Answer: 0.12389366636436545

#### 5.  What is the probability that the walker is at node 6 after 50 steps?

In [14]:
fiftied = np.linalg.matrix_power(B, 50)

In [15]:
fiftied[0,5]

0.09078488881507354

Answer: 0.09078488881507354

#### 6.  What is the probability that the walker is at node 2 after 50 steps?

In [16]:
fiftied[0,1]

0.09101798714914272

Answer: 0.09101798714914272

### Part 2:

Make a function findk(n,p,error) and answer some questions.

In [17]:
def findk(n,p,error):
    """
    Determines the exponent k that will most closely satisfy our desired n (nodes), p (probability), and error.
    
    Recall:
            max{||(S**k * x) - q|| : x is a probabilty vector} =< sqrt((n - 1) * |lambda2 * k|)
            q = (1/n)*np.ones((n,1)), 0 < p < 1
            S,L = rw_matrix(n,p) ## S is the matrix, L is the lambda2
    
    Inputs: Our desired n (nodes), p (probability), and error
    Output: an integer k satisfying:
            max{||(S**k * x) - q|| : x is a probabilty vector}
    """
    q = (1/n)*np.ones((n,1))
    S,L = rw_matrix(n,p)
    
    k = 1 
    while True:
        if np.sqrt((n - 1) * (np.abs(L)**k)) < error:
            return k  
        k = k + 1  

#### 7. For what value of k is within 0.01 of (1/15)*np.ones((15,1))?

In [18]:
findk(15,.25,.01)

177

Answer: k = 177

#### 8. For what value of k is within 0.001 of (1/15)*np.ones((15,1))?

In [19]:
findk(15,.25,.001)

246

Answer: k = 246

#### 9. For what value of k is within 0.01 of (1/150)*np.ones((150,1))?

In [20]:
findk(150,.25,.01)

21600

Answer: k = 21600

#### 10. For what value of k is within 0.001 of (1/150)*np.ones((150,1))?

In [21]:
findk(150,.25,.001)

28597

Answer: k = 28597

#### 11. Check your answers with the probability vector x = [1,0,...,0].T in 7-10 above (no markdown cell needed for this one)

In [22]:
def check_k(n,p,error,k):
    """
    Ensures the k found in findk works in ||(S**k * x) - q||, when x = [1,0,...,0].T by returning a Boolean value
    Inputs: Our desired n (nodes), p (probability), error, and k guess
    Output: A boolean value
    """
    q = (1/n)*np.ones((n,1))
    S,L = rw_matrix(n,p)
    
    x = np.zeros((n,1))
    x[0] = 1
    
    result = np.dot(np.linalg.matrix_power(S, k), x) - q
    new_result = np.linalg.norm(result)
    return print(new_result < error)

In [23]:
check_k(15,.25,.01,177)
check_k(15,.25,.001,246)
check_k(150,.25,.01,21600)
check_k(150,.25,.001,28597)

True
True
True
True
